contracts/Aavegotchi/facets/ERC721BuyOrderFacet.sol

```diff
@@ -0,0 +1,190 @@
+
+        s.erc721BuyOrders[buyOrderId] = ERC721BuyOrder({
+            buyOrderId: buyOrderId,
+            buyer: sender,
+            erc721TokenAddress: _erc721TokenAddress,
+            erc721TokenId: _erc721TokenId,
+            priceInWei: _priceInWei,
+            validationHash: LibBuyOrder.generateValidationHash(_erc721TokenAddress, _erc721TokenId, _validationOptions),
+            timeCreated: block.timestamp,
+            timePurchased: 0,
+            cancelled: false,
+            validationOptions: _validationOptions
+        });
+        emit ERC721BuyOrderAdd(buyOrderId, sender, _erc721TokenAddress, _erc721TokenId, category, _priceInWei, block.timestamp);
+    }
+
+    function cancelERC721BuyOrder(uint256 _buyOrderId) external {
+        address sender = LibMeta.msgSender();
+        ERC721BuyOrder memory erc721BuyOrder = s.erc721BuyOrders[_buyOrderId];
+
+        require(erc721BuyOrder.timeCreated != 0, "ERC721BuyOrder: ERC721 buyOrder does not exist");
+        require(
+            (sender == s.aavegotchis[erc721BuyOrder.erc721TokenId].owner) || (sender == erc721BuyOrder.buyer),
+            "ERC721BuyOrder: Only aavegotchi owner or buyer can call this function"
+        );
+        require((erc721BuyOrder.cancelled == false) && (erc721BuyOrder.timePurchased == 0), "ERC721BuyOrder: Already processed");
+
+        LibBuyOrder.cancelERC721BuyOrder(_buyOrderId);
+    }
+
+    function executeERC721BuyOrder(uint256 _buyOrderId) external {
+        address sender = LibMeta.msgSender();
+        ERC721BuyOrder memory erc721BuyOrder = s.erc721BuyOrders[_buyOrderId];
+
+        require(erc721BuyOrder.timeCreated != 0, "ERC721BuyOrder: ERC721 buyOrder does not exist");
+        require(sender == s.aavegotchis[erc721BuyOrder.erc721TokenId].owner, "ERC721BuyOrder: Only aavegotchi owner can call this function");
+        require((erc721BuyOrder.cancelled == false) && (erc721BuyOrder.timePurchased == 0), "ERC721BuyOrder: Already processed");
+
+        // disable for gotchi in lending
+        uint256 category = LibAavegotchi.getERC721Category(erc721BuyOrder.erc721TokenAddress, erc721BuyOrder.erc721TokenId);
+        if (category == LibAavegotchi.STATUS_AAVEGOTCHI) {
+            uint32 listingId = s.aavegotchiToListingId[uint32(erc721BuyOrder.erc721TokenId)];
+            require((listingId == 0) || (s.gotchiLendings[listingId].timeAgreed == 0), "ERC721BuyOrder: Not supported for aavegotchi in lending");
+        }
+
+        // hash validation
+        require(
+            erc721BuyOrder.validationHash ==
+                LibBuyOrder.generateValidationHash(erc721BuyOrder.erc721TokenAddress, erc721BuyOrder.erc721TokenId, erc721BuyOrder.validationOptions),
+            "ERC721BuyOrder: Invalid buy order"
+        );
+
+        erc721BuyOrder.timePurchased = block.timestamp;
+
+        LibERC721Marketplace.cancelERC721Listing(erc721BuyOrder.erc721TokenAddress, erc721BuyOrder.erc721TokenId, sender);
```

Comment:
This should already happen when the NFT is transferred right? Do we need this code?

Reply:
Yes, this is needed also.

Reply:
Can you explain the reason please @orionstardust

Reply:
In gotchi transfer function, there's no code for removing listing.
They are separated. So we need this.

Reply:
`internalTransferFrom` contains this function on Line 241 

`LibERC721Marketplace.updateERC721Listing(address(this), _tokenId, _from);`

Reply:
oh, I missed it. I just updated.